File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxProduct(TreeNode root) {
        long sum = findSum(root);
        long[] res = new long[]{sum, 0};
        findMaxProduct(root, res);
        return (int) (res[1] % (1e9 + 7));
    }
    
    public long findMaxProduct(TreeNode root, long[] res){
        if(root == null)
            return 0;
        
        long left = findMaxProduct(root.left, res);
        long right = findMaxProduct(root.right, res);
        long curr = left + root.val + right;
        res[1] = Math.max(res[1], curr * (res[0] - curr));
        return curr;
    }
    
    public long findSum(TreeNode root){
        if(root == null)
            return 0;
        
        return findSum(root.left) + root.val + findSum(root.right);
    }
}

File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def find_sum(self, node):
        if node == None:
            return 0
        
        return self.find_sum(node.left) + node.val + self.find_sum(node.right)
    
    def find_max_product(self, node, res):
        if node == None:
            return 0
        
        left = self.find_max_product(node.left, res)
        right = self.find_max_product(node.right, res)
        curr = left + node.val + right
        res[1] = max(res[1], (res[0] - curr) * curr)
        return curr
    
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        if root == None:
            return 0
        
        total = self.find_sum(root)
        res = [total, 0]
        self.find_max_product(root, res)
        return (int)(res[1] % (1e9 + 7))
        
        